### PR TITLE
Issue 5982: Fix race in TcpClientConnection.

### LINKED ISSUE #5982: Network glitch causes connectionDropped to be invoked on a writer post which a SegmentSealed response is processed.
**Describe the bug**
It was observed that a ScalingRetransmission thread was blocked in our medium scale longevity run.
```
ScalingRetransmission-longevity/mediumScale-4075" #86380 daemon prio=5 os_prio=0 tid=0x00007f17b0786800 nid=0x555b waiting for monitor entry [0x00007f16ca130000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$handleLogSealed$2(EventStreamWriterImpl.java:182)
	- waiting to lock <0x00000000dc714dc8> (a java.lang.Object)
	at io.pravega.client.stream.impl.EventStreamWriterImpl$$Lambda$459/1509050028.attempt(Unknown Source)
	at io.pravega.common.util.Retry$RetryAndThrowBase.run(Retry.java:173)
	at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$handleLogSealed$3(EventStreamWriterImpl.java:173)
```
The thread is waiting for a lock `0x00000000dc714dc8` . This lock is acquired by a different thread which is attempting to write the next event to the Pravega stream.

```
"pool-6-thread-3" #500 prio=5 os_prio=0 tid=0x00007f17f0bbb800 nid=0x20c waiting on condition [0x00007f16c19ad000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000e742de78> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
	at io.pravega.common.concurrent.Futures.getThrowingException(Futures.java:178)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl.write(SegmentOutputStreamImpl.java:490)
	- locked <0x00000000e51f5f30> (a java.lang.Object)
	at io.pravega.client.stream.impl.EventStreamWriterImpl.writeEventInternal(EventStreamWriterImpl.java:125)
	- locked <0x00000000dc714dc8> (a java.lang.Object)
```
This is triggered because the writer does not expect any response from the SSS post a connectin drop.
```
2021-04-19 04:37:24,571 307495455 [pool-6-thread-3] WARN  i.p.c.s.i.SegmentOutputStreamImpl - Failed writing event through writer 70132054-ff8a-4d21-ac90-13410bd057d0 due to: 
io.pravega.shared.protocol.netty.ConnectionFailedException: java.net.SocketException: Broken pipe (Write failed)
	at io.pravega.client.connection.impl.TcpClientConnection.send(TcpClientConnection.java:289)
	at io.pravega.client.connection.impl.FlowClientConnection.send(FlowClientConnection.java:48)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl.write(SegmentOutputStreamImpl.java:504)
2021-04-19 04:37:24,571 307495455 [ClientSocketReaders-53708] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Received SegmentSealed WireCommands.SegmentIsSealed(type=SEGMENT_IS_SEALED, requestId=302820964171776, segment=longevity/mediumScale/64056.#epoch.41917, serverStackTrace=, offset=8971) on writer 70132054-ff8a-4d21-ac90-13410bd057d0
```
**To Reproduce**
WireCommands.SegmentIsSealed is received post a connection drop.

**Screenshots**
N/A

**Additional information**
Fix the underlying issue instead of only fixing it on the writer.

Details logs:
[networkBugLogs.zip](https://github.com/pravega/pravega/files/6399592/networkBugLogs.zip)

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
Fix a race in TcpClientConnection. This ensures `ReplyProcessor.connectionDropped()` is only invoked after `ConnectionReader` has stopped reading from the socket.

**Purpose of the change**  
Fixes #5982 

**What the code does**  
This ensures `ReplyProcessor.connectionDropped()` is only invoked after `ConnectionReader ` has stopped reading from the `TCP` socket. Without this change, it can so happen that the other Replyprocessor callbacks (like ReplyProcessor.segmentIsSealed(..)) is invoked post a callback indicating a `connectionDrop`.  
All the clients do not expect any message after a connection drop. 

**How to verify it**  
All the existing tests should pass.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java [hunk 1]
```diff
@@ -117,31 +121,52 @@ public void run() {
             IoBuffer buffer = new IoBuffer();
             while (!stop.get()) {
                 try {
+                    isReadingFromSocket.set(true);
+                    // This method blocks until it is able to read data from the tcp socket InputStream.
                     WireCommand command = readCommand(in, buffer);
+                    isReadingFromSocket.set(false);
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:127]: What happens if the line above throws? Should we reset this value to `false` in a `finally` block?

**Comment by shrids** [comment]
System test run with this changeset is successful `job/test-pravega-continuous/3804/`

File: client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java (hunk 1 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:127]: An exception from `readCommand(..)` method will be caught and the Socket inputStream will be closed . Also, a concurrent `stop()` invocation will be a NOP. Hence I do not see a need for the finally block here.

File: client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java [hunk 2]
```diff
@@ -168,6 +193,11 @@ public void stop() {
             if (stop.getAndSet(true)) {
                 return;
             }
+            if (!isReadingFromSocket.get() && !isCallbackExecuting.get()) {
+                // wait until we have completed the current call to the reply processors.
+                Exceptions.handleInterrupted(hasStopped::await);
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:198]: If this were moved to after the `closeQuietly` call below, couldn't we avoid the need for these two atomic bools?

Comment (shrids) [reply client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:188]: Thanks, I have simplified the code.

Comment (shrids) [reply client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:184]: Thanks. I have updated the fix . The code also needs to handle scenarios where stop() is invoked by `ClientSocketReaders` thread.

**Review by tkaitchuck — APPROVED** [review]

**Review by tkaitchuck — APPROVED** [review]

**Pull request merged**